src/Arcus.Messaging.Pumps.ServiceBus/ServiceBusMessageRouter.cs

```diff
@@ -0,0 +1,197 @@
+                }
+
+                await DeadLetterMessageNoHandlerMatchedAsync(messageContext);
+                return MessageProcessingResult.Failure(message.MessageId, CannotFindMatchedHandler, "Failed to process Azure Service Bus message in pump as no message handler was matched against the message");
+            }
+            catch (Exception exception)
+            {
+                Logger.LogCritical(exception, "Unable to process message with ID '{MessageId}'", message.MessageId);
+
+                await messageContext.AbandonMessageAsync(new Dictionary<string, object>(), CancellationToken.None);
+                return MessageProcessingResult.Failure(message.MessageId, ProcessingInterrupted, "Failed to process message in pump as there was an unexpected critical problem during processing, please see the logs for more information", exception);
+            }
+        }
+
+        private static string LoadMessageBody(ServiceBusReceivedMessage message, AzureServiceBusMessageContext context)
+        {
+            Encoding encoding = DetermineEncoding();
+            string messageBody = encoding.GetString(message.Body.ToArray());
+
+            return messageBody;
+
+            Encoding DetermineEncoding()
+            {
+                Encoding fallbackEncoding = Encoding.UTF8;
+
+                if (context.Properties.TryGetValue(PropertyNames.Encoding, out object encodingNameObj)
+                    && encodingNameObj is string encodingName
+                    && !string.IsNullOrWhiteSpace(encodingName))
+                {
+                    EncodingInfo foundEncoding =
+                        Encoding.GetEncodings()
+                                .FirstOrDefault(e => e.Name.Equals(encodingName, StringComparison.OrdinalIgnoreCase));
+
+                    return foundEncoding?.GetEncoding() ?? fallbackEncoding;
+                }
+
+                return fallbackEncoding;
+            }
+        }
+
+        private async Task PotentiallyAutoCompleteMessageAsync(AzureServiceBusMessageContext messageContext)
+        {
+            if (_serviceBusOptions.AutoComplete)
+            {
+                try
+                {
+                    Logger.LogTrace("Auto-complete message '{MessageId}' (if needed) after processing in Azure Service Bus in message pump '{JobId}'", messageContext.MessageId, messageContext.JobId);
+                    await messageContext.CompleteMessageAsync(CancellationToken.None);
+                }
+                catch (ServiceBusException exception) when (
+                    exception.Message.Contains("lock")
+                    && exception.Message.Contains("expired")
+                    && exception.Message.Contains("already")
+                    && exception.Message.Contains("removed"))
```

Comment:
I think it's a bit 'dangerous' to filter on (sub)strings that are part of the exception-message.
I believe we should have a look if we can't make use of the `Reason` property of the `ServiceBusException` class.

Reply:
It was done in the session pump, but not in the receiver pump.